src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -1240,5 +1240,154 @@ private static bool HasNamespaceName(NamespaceSymbol @namespace, string namespac
             var name = @namespace.Name;
             return (name.Length == length) && (string.Compare(name, 0, namespaceName, offset, length, comparison) == 0);
         }
+
+        internal static bool IsNonGenericTaskType(this TypeSymbol type, CSharpCompilation compilation)
+        {
+            var namedType = type as NamedTypeSymbol;
+            if ((object)namedType == null || namedType.Arity != 0)
+            {
+                return false;
+            }
+            if ((object)namedType == compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task))
+            {
+                return true;
+            }
+            if (namedType.SpecialType == SpecialType.System_Void)
+            {
+                return false;
+            }
+            MethodSymbol createBuilderMethod;
+            var builderType = namedType.GetAsyncMethodBuilderType(out createBuilderMethod);
+            return (object)builderType != null;
+        }
+
+        internal static bool IsGenericTaskType(this TypeSymbol type, CSharpCompilation compilation)
+        {
+            var namedType = type as NamedTypeSymbol;
+            if ((object)namedType == null || namedType.Arity != 1)
+            {
+                return false;
+            }
+            if ((object)namedType.ConstructedFrom == compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task_T))
+            {
+                return true;
+            }
+            MethodSymbol createBuilderMethod;
+            var builderType = namedType.GetAsyncMethodBuilderType(out createBuilderMethod);
+            return (object)builderType != null;
+        }
+
+        /// <summary>
+        /// Returns the async method builder type for generic and non-generic task-like types.
+        /// </summary>
+        internal static NamedTypeSymbol GetAsyncMethodBuilderType(this NamedTypeSymbol type, out MethodSymbol createBuilderMethod)
+        {
+            Debug.Assert((object)type != null);
+            Debug.Assert(type.SpecialType != SpecialType.System_Void);
```

Comment:
Changed the name and signature of the method to your suggestion and checked arity.